### DISCUSSION TITLE
Enhanced sleep mode

### DISCUSSION THREAD

**Discussion by patrickwelker** [LeoKlaus/plappa#103] [Feature Requests]
- an option in the settings for the sleep timer to be __always enabled__. My favorite audiobook app introduced this some time ago and I happen to find it brilliant. You open the app and no matter what, the audiobook will play only ten minutes (i. e. my set duration for the sleep timer). This is a life saver for people who use audiobooks to fall asleep when they (still) awake. If I listen to audiobooks while cooking, coding or coloring books I simply deactivate the timer.
- shake to extend sleep timer should also work after the timer runs out since that is usually the moment when one realizes „okay, I’m still awake, let’s give it another ten minutes“ or simply „okay, one more round“. _(Most sleep timers have a grace period of ~5-15 minutes to accept/detect the shake… probably restricted by how background apps work on  iOS)

I attached some screenshots from the settings and dedicated sleep timer settings of [MP3 Audiobook Player](https://apps.apple.com/de/app/mp3-audiobook-player/id891797540?l=en-GB) for reference.

![image](https://github.com/LeoKlaus/plappa/assets/591759/b061fcc4-f428-41bc-869d-b2aecccd38de) ![image](https://github.com/LeoKlaus/plappa/assets/591759/0fd7c24f-fc73-4384-b4ee-5e40822d4128)

Edit: The shake to wake is working with locked phones, too. I just tested it again by setting my iPhone to lock immediately (instead of 5 min) and setting a one minute sleep timer.

The shake “restarts” the app and continues play. Maybe that’s the workaround: a restart aka “play last book (/w new timer)”.

**Comment by LeoKlaus** [comment]
Hey Patrick,

thank you for taking the time to post this on GitHub. I'll definitely have a look into this when I find the time!

**Reply by patrickwelker** [reply]
Since I find the shake to extend (even with locked phone) crucial in an audiobook app. I haven't continued using Plappa.

However, I asked perplexity.ai and maybe these findings and the relevant sources help. It's not the best resources, but one can hope:

To implement a sleep timer feature with an "always on" option and a "shake to extend" functionality in an iPhone app, you can follow these code examples using Swift and UIKit.

### Setting Up the Sleep Timer

1. **Create a Timer Variable**:
   Define a variable to hold your timer.

   ```swift
   var sleepTimer: Timer?
   ```

2. **Start the Timer**:
   Create a function to start the timer with a specified duration.

   ```swift
   func startSleepTimer(duration: TimeInterval) {
       sleepTimer?.invalidate() // Invalidate any existing timer
       sleepTimer = Timer.scheduledTimer(withTimeInterval: duration, repeats: false) { [weak self] _ in
           self?.stopPlayback()
       }
   }
   ```

3. **Stop Playback**:
   Implement a function to stop the audio playback when the timer expires.

   ```swift
   func stopPlayback() {
       // Code to stop audio playback
       print("Playback stopped")
   }
   ```

### Enabling Always-On Timer

4. **Persistent Sleep Timer**:
   Use `UserDefaults` to remember the user's preference for an always-on timer.

   ```swift
   func setAlwaysOnSleepTimer(enabled: Bool) {
       UserDefaults.standard.set(enabled, forKey: "alwaysOnSleepTimer")
       if enabled {
           startSleepTimer(duration: 600) // Set to 10 minutes
       }
   }

   func checkAlwaysOnSleepTimer() {
       if UserDefaults.standard.bool(forKey: "alwaysOnSleepTimer") {
           startSleepTimer(duration: 600)
       }
   }
   ```

### Implementing Shake to Extend

5. **Detect Shake Gesture**:
   Override the motion detection method to handle shake gestures.

   ```swift
   override func motionEnded(_ motion: UIEvent.EventSubtype, with event: UIEvent?) {
       if motion == .motionShake {
           extendSleepTimer()
       }
   }

   func extendSleepTimer() {
       // Restart the timer for an additional duration
       startSleepTimer(duration: 600) // Extend by 10 minutes
       print("Sleep timer extended")
   }
   ```

### Usage

- Call `checkAlwaysOnSleepTimer()` when the app loads or when starting playback to ensure the timer is set based on user preferences.
- Use `setAlwaysOnSleepTimer(enabled:)` to toggle the always-on feature in your app's settings.

This setup allows users to have an audiobook play for a set duration automatically and extend it easily by shaking their device, enhancing their listening experience while promoting relaxation.

Sources
[1] [iOS] Auto sleep timer · Issue #260 · advplyr/audiobookshelf-app https://github.com/advplyr/audiobookshelf-app/issues/260
[2] Shake to extend sleep timer there and gone? : r/audible - Reddit https://www.reddit.com/r/audible/comments/z6qtco/shake_to_extend_sleep_timer_there_and_gone/
[3] Does anyone else have this problem with Audible's sleep timer ... https://www.reddit.com/r/audiobooks/comments/65wqxs/does_anyone_else_have_this_problem_with_audibles/
[4] Audible Sleep Timer Hacks: Setup, Fixes & Customization - ViWizard https://www.viwizard.com/audiobook-tips/set-sleep-timer-audible.html
[5] Looking for iOS audiobook app with "shake" function https://www.mobileread.com/forums/showthread.php?p=3694765
[6] Audible: Audio Entertainment on the App Store https://apps.apple.com/us/app/audible-audio-entertainment/id379693831
[7] iphone audiobook player with sleep timer on by default? - Reddit https://www.reddit.com/r/audiobooks/comments/ukb6k9/iphone_audiobook_player_with_sleep_timer_on_by/
[8] Better sleep timer with shake to extend #4580 - GitHub https://github.com/AntennaPod/AntennaPod/issues/4580
